File: scripts/preprocessLocal.js
#!/usr/bin/env node
/**
 * Preprocesses the local Baltimore Real Property GeoJSON into a browser-ready
 * subset (commercial/industrial/notable parcels only).
 *
 * Input:  data/Real_Property_Information.geojson   (563 MB, all parcels)
 * Output: public/data/commercial-properties.geojson (~5–10 MB, filtered)
 *
 * Run: node scripts/preprocessLocal.js
 */

import { createReadStream, createWriteStream, mkdirSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const IN_FILE  = join(__dirname, '../data/Real_Property_Information.geojson');
const OUT_DIR  = join(__dirname, '../public/data');
const OUT_FILE = join(OUT_DIR, 'commercial-properties.geojson');

const COMMERCIAL_USEGROUPS = new Set(['C', 'I', 'M', 'CC', 'OR']);
const MIN_VALUE = 500_000;

function keepFeature(props) {
  const ug = (props.USEGROUP || '').trim().toUpperCase();
  const value = props.ARTAXBAS || props.FULLCASH || props.TAXBASE || 0;
  return COMMERCIAL_USEGROUPS.has(ug) || value >= MIN_VALUE;
}

function slimProps(p) {
  return {
    OBJECTID:   p.OBJECTID,
    BLOCKLOT:   p.BLOCKLOT,
    FULLADDR:   p.FULLADDR,
    OWNER_1:    p.OWNER_1,
    OWNER_2:    p.OWNER_2,
    USEGROUP:   p.USEGROUP,
    ZONECODE:   p.ZONECODE,
    ARTAXBAS:   p.ARTAXBAS,
    TAXBASE:    p.TAXBASE,
    CURRLAND:   p.CURRLAND,
    CURRIMPR:   p.CURRIMPR,
    LOT_SIZE:   p.LOT_SIZE,
    YEAR_BUILD: p.YEAR_BUILD,
    STRUCTAREA: p.STRUCTAREA,
    SALEPRIC:   p.SALEPRIC,
    SALEDATE:   p.SALEDATE,
    NEIGHBOR:   p.NEIGHBOR,
    ZIP_CODE:   p.ZIP_CODE,
    SDATLINK:   p.SDATLINK,
    DHCDUSE1:   p.DHCDUSE1,
  };
}

console.log('\n  Baltimore Real Property Preprocessor');
console.log('='.repeat(50));
console.log('Stream-parsing 563 MB GeoJSON…\n');

// Stream parse strategy: read file in chunks, accumulate until we have a full
// feature JSON object, then parse and filter it.
// GeoJSON feature collections look like:
//   {"type":"FeatureCollection","features":[{...},{...},...]}
// Each feature is separated by "},\n{" or similar.
// We buffer until we find balanced braces for each feature.

mkdirSync(OUT_DIR, { recursive: true });

const outStream = createWriteStream(OUT_FILE);
outStream.write('{"type":"FeatureCollection","generatedAt":"' +
  new Date().toISOString() + '","sourceFile":"Real_Property_Information.geojson","features":[');

let buffer = '';
let inFeatures = false;
let braceDepth = 0;
let featureStart = -1;
let total = 0;
let kept = 0;
let firstOut = true;

const stream = createReadStream(IN_FILE, { encoding: 'utf8', highWaterMark: 1 << 20 }); // 1MB chunks

stream.on('data', chunk => {
  buffer += chunk;

  // Scan for feature boundaries
  let i = 0;

  if (!inFeatures) {
    // Find the start of the features array
    const idx = buffer.indexOf('"features"');
    if (idx === -1) { buffer = buffer.slice(-20); return; }
    const arrStart = buffer.indexOf('[', idx);
    if (arrStart === -1) { buffer = buffer.slice(-(buffer.length - idx)); return; }
    buffer = buffer.slice(arrStart + 1);
    inFeatures = true;
    i = 0;
  }

  while (i < buffer.length) {
    const ch = buffer[i];
    if (featureStart === -1) {
      if (ch === '{') { featureStart = i; braceDepth = 1; }
      i++;
      continue;
    }
    if (ch === '{') braceDepth++;
    else if (ch === '}') {
      braceDepth--;
      if (braceDepth === 0) {
        // Complete feature found
        const raw = buffer.slice(featureStart, i + 1);
        try {
          const f = JSON.parse(raw);
          total++;
          if (total % 10000 === 0) process.stdout.write(`\r  Scanned ${total.toLocaleString()} features, kept ${kept.toLocaleString()}…`);
          if (f.geometry &&
              (f.geometry.type === 'Polygon' || f.geometry.type === 'MultiPolygon') &&
              keepFeature(f.properties || {})) {
            const out = JSON.stringify({ type: 'Feature', geometry: f.geometry, properties: slimProps(f.properties) });
            outStream.write((firstOut ? '' : ',') + out);
            firstOut = false;
            kept++;
          }
        } catch (_) { /* malformed feature, skip */ }
        buffer = buffer.slice(i + 1);
        i = 0;
        featureStart = -1;
        continue;
      }
    }
    i++;
  }

  // Keep only the unprocessed tail
  if (featureStart !== -1) {
    buffer = buffer.slice(featureStart);
    featureStart = 0;
  } else {
    buffer = buffer.slice(-2); // keep last 2 chars in case brace straddles chunk
    featureStart = -1;
  }
});

stream.on('end', () => {
  outStream.write(']}');
  outStream.end(() => {
    process.stdout.write('\r' + ' '.repeat(70) + '\r');
    console.log(`\nTotal scanned:   ${total.toLocaleString()}`);
    console.log(`Kept (filtered): ${kept.toLocaleString()}`);
    console.log(`\n  Written: ${OUT_FILE}`);
    console.log('  Restart npm run dev — app will load this file automatically.\n');
  });
});

stream.on('error', e => { console.error('Read error:', e); process.exit(1); });

File: scripts/preprocessLocal.py
#!/usr/bin/env python3
"""
Preprocesses Baltimore Real Property GeoJSON into a browser-ready subset.

Input:  data/Real_Property_Information.geojson   (563 MB)
Output: public/data/commercial-properties.geojson

Run: python scripts/preprocessLocal.py
"""

import json, sys, os
from pathlib import Path

ROOT     = Path(__file__).parent.parent
IN_FILE  = ROOT / "data" / "Real_Property_Information.geojson"
OUT_DIR  = ROOT / "public" / "data"
OUT_FILE = OUT_DIR / "commercial-properties.geojson"

COMMERCIAL_USEGROUPS = {"C", "I", "M", "CC", "OR"}
MIN_VALUE = 500_000

KEEP_PROPS = [
    "OBJECTID", "BLOCKLOT", "FULLADDR", "OWNER_1", "OWNER_2",
    "USEGROUP", "ZONECODE", "ARTAXBAS", "TAXBASE", "CURRLAND", "CURRIMPR",
    "LOT_SIZE", "YEAR_BUILD", "STRUCTAREA", "SALEPRIC", "SALEDATE",
    "NEIGHBOR", "ZIP_CODE", "SDATLINK", "DHCDUSE1",
]

def keep(props):
    ug    = (props.get("USEGROUP") or "").strip().upper()
    value = props.get("ARTAXBAS") or props.get("FULLCASH") or props.get("TAXBASE") or 0
    return ug in COMMERCIAL_USEGROUPS or value >= MIN_VALUE

def slim(props):
    return {k: props.get(k) for k in KEEP_PROPS}

print("\n  Baltimore Real Property Preprocessor")
print("=" * 50)
print(f"Input:  {IN_FILE}")
print(f"Output: {OUT_FILE}\n")

if not IN_FILE.exists():
    sys.exit(f"ERROR: {IN_FILE} not found")

OUT_DIR.mkdir(parents=True, exist_ok=True)

try:
    import ijson
    USE_IJSON = True
except ImportError:
    USE_IJSON = False

if USE_IJSON:
    print("Streaming with ijson…")
    total = kept = 0
    features_out = []

    with open(IN_FILE, "rb") as f:
        for feature in ijson.items(f, "features.item"):
            total += 1
            if total % 10_000 == 0:
                print(f"\r  Scanned {total:,}  kept {kept:,}…", end="", flush=True)
            geom = feature.get("geometry")
            if not geom:
                continue
            if geom.get("type") not in ("Polygon", "MultiPolygon"):
                continue
            props = feature.get("properties") or {}
            if not keep(props):
                continue
            features_out.append({"type": "Feature", "geometry": geom, "properties": slim(props)})
            kept += 1

else:
    # Fallback: line-by-line scan — works because each feature is on one line
    # in Baltimore's GeoJSON export.
    print("ijson not available, using line-by-line fallback…")
    total = kept = 0
    features_out = []

    with open(IN_FILE, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip().rstrip(",")
            if not line.startswith('{"type":"Feature"') and not line.startswith('{ "type": "Feature"'):
                continue
            total += 1
            if total % 10_000 == 0:
                print(f"\r  Scanned {total:,}  kept {kept:,}…", end="", flush=True)
            try:
                feature = json.loads(line)
            except json.JSONDecodeError:
                continue
            geom = feature.get("geometry")
            if not geom or geom.get("type") not in ("Polygon", "MultiPolygon"):
                continue
            props = feature.get("properties") or {}
            if not keep(props):
                continue
            features_out.append({"type": "Feature", "geometry": geom, "properties": slim(props)})
            kept += 1

print(f"\r{' ' * 60}\r", end="")
print(f"Scanned:  {total:,}")
print(f"Kept:     {kept:,}")

output = {
    "type": "FeatureCollection",
    "generatedAt": __import__("datetime").datetime.now(__import__("datetime").timezone.utc).isoformat(),
    "sourceFile": "Real_Property_Information.geojson",
    "features": features_out,
}

import decimal

class DecimalEncoder(json.JSONEncoder):
    def default(self, o):
        if isinstance(o, decimal.Decimal):
            return float(o)
        return super().default(o)

with open(OUT_FILE, "w") as f:
    json.dump(output, f, separators=(",", ":"), cls=DecimalEncoder)

size_mb = OUT_FILE.stat().st_size / 1024 / 1024
print(f"\n  Written: {OUT_FILE}  ({size_mb:.1f} MB)")
print("  Restart npm run dev — the app will load this file automatically.\n")
